README.md
# citygml2-to-citygml3

A Java tool used to transform CityGML v2.0 to CityGML v3.0 files.

This project is still in early development and subject to change. (However, it is already usable/useful and many datasets from different cities and countries could be successfully transformed.) 

### How to run

1. The source code can be cloned using 
   ```batch
   git clone https://github.com/tum-gis/citygml2-to-citygml3
   ```
   or downloading the latest release [here](https://github.com/tum-gis/citygml2-to-citygml3/releases).

2. The program can then be executed by running the `Run.bat` file (in Windows) or using the following command line:
   ```batch
   java -jar citygml2-to-citygml3.jar
   ```

3. The result is by default saved in folder [output](output), such as [output/FZKHaus_v3.gml](output/FZKHaus_v3.gml) (only visible after the program is complete). 
   Simply open this folder/file to see the output.

### Configurations
The program settings can be adjusted in the configuration file [SETTINGS.txt](SETTINGS.txt). Depending on the use-case, users may change these configurations to fit their needs:

| Parameter        | Description           | Default  |
| ------------- |:-------------| -----|
| `INPUT_FILE`      | The location of the source CityGML v2.0 file | `input/FZKHaus.gml` |
| `XSL_FILE`      | The location of the XSL file used for transformation | `xsl/Transform.xsl` |
| `OUTPUT_FILE`      | The location of the transformed CityGML v3.0 file | `output/FZKHaus_v3.gml` |
| `LOD4_TO_LOD3`      | Set to `true` if all found LOD 4 elements are to be replaced by the corresponding LOD 3 elements. The value `false` will remove all found LOD 4 elements. | `true` |
| `LOD4_GEOMETRY`      | Specifies how to handle found `lod4Geometry` objects. If the `LOD4_TO_LOD3` is set to `true`, `lod4Geometry` can either be converted to `lod3MultiSurface`, `lod3MultiCurve` or`lod3Solid`. | `lod3MultiSurface` |

For reference, the file [FZKHaus_v3.gml](output/FZKHaus_v3.gml), which contains all LOD 0-4 information, can be used as an example of how the transformed file should look like.

### Namespaces
To produce the best results, the input CityGML document must satisfy the following conditions:

1. It must be encoded in CityGML v2.0;

2. All namespaces and schemata must be declared in the root element (i.e. `CityModel`) and nowhere else (e.g. local namespaces in elements other than the root are not recommended);

3. Namespace prefixes and URIs as well as schema locations must correspond to the values allowed by the tool. 

Please ensure to check the namespaces and schemata of your CityGML documents accordingly before executing the tool.

The list of allowed namespaces in CityGML v2.0 as well as in v3.0 are selected by default as follows:

| Namespace prefix        | Module |       Namespace URI in CityGML v2.0           |      Namespace URI in CityGML v3.0  |
| ------------- |------|-------------| -----|
| `xmlns:app` | Appearance | `http://www.opengis.net/citygml/appearance/2.0` | `http://www.opengis.net/citygml/appearance/3.0` |
| `xmlns:brid` | Bridge | `http://www.opengis.net/citygml/bridge/2.0` | `http://www.opengis.net/citygml/bridge/3.0` |
| `xmlns:bldg` | Building | `http://www.opengis.net/citygml/building/2.0` | `http://www.opengis.net/citygml/building/3.0` |
| `xmlns:frn` | City Furniture | `http://www.opengis.net/citygml/cityfurniture/2.0` | `http://www.opengis.net/citygml/cityfurniture/3.0` |
| `xmlns:grp` | City Object Group | `http://www.opengis.net/citygml/cityobjectgroup/2.0` | `http://www.opengis.net/citygml/cityobjectgroup/3.0` |
| **`xmlns:con`** | **Construction** |  | **`http://www.opengis.net/citygml/construction/3.0`** |
| `xmlns:core` | Core | `http://www.opengis.net/citygml/2.0` | `http://www.opengis.net/citygml/3.0` |
| **`xmlns:dyn`** | **Dynamizer** |  | **`http://www.opengis.net/citygml/dynamizer/3.0`** |
| `xmlns:gen` | Generics | `http://www.opengis.net/citygml/generics/2.0` | `http://www.opengis.net/citygml/generics/3.0` |
| `xmlns:luse` | LandUse | `http://www.opengis.net/citygml/landuse/2.0` | `http://www.opengis.net/citygml/landuse/3.0` |
| `xmlns:dem` | Relief | `http://www.opengis.net/citygml/relief/2.0` | `http://www.opengis.net/citygml/relief/3.0` |
| **`xmlns:pcl`** | **PointCloud** |  | **`http://www.opengis.net/citygml/pointcloud/3.0`** |
| `xmlns:tran` | Transportation | `http://www.opengis.net/citygml/transportation/2.0` | `http://www.opengis.net/citygml/transportation/3.0` |
| `xmlns:tun` | Tunnel | `http://www.opengis.net/citygml/tunnel/2.0` | `http://www.opengis.net/citygml/tunnel/3.0` |
| `xmlns:veg` | Vegetation | `http://www.opengis.net/citygml/vegetation/2.0` | `http://www.opengis.net/citygml/vegetation/3.0` |
| **`xmlns:vers`** | **Versioning** |  | **`http://www.opengis.net/citygml/versioning/3.0`** |
| `xmlns:wtr` | WaterBody | `http://www.opengis.net/citygml/waterbody/2.0` | `http://www.opengis.net/citygml/waterbody/3.0` |
| **`xmlns:tsml`** | **TimeseriesML** |  | **`http://www.opengis.net/tsml/1.0`** |
| `xmlns:sos` | Sensor Observation Service | `http://www.opengis.net/sos/2.0` | `http://www.opengis.net/sos/2.0` |
| `xmlns:xAL` | eXtensible Address Language | `urn:oasis:names:tc:ciq:xsdschema:xAL:2.0` | `urn:oasis:names:tc:ciq:xsdschema:xAL:2.0` |
| `xmlns:xlink` | XLink | `http://www.w3.org/1999/xlink` | `http://www.w3.org/1999/xlink` |
| `xmlns:xsi` | XML Schema Instance | `http://www.w3.org/2001/XMLSchema-instance` | `http://www.w3.org/2001/XMLSchema-instance` |
| `xmlns:gml` | Geography Markup Language | `http://www.opengis.net/gml` | `http://www.opengis.net/gml/3.2` |
| `xmlns` | Default namespace | `http://www.opengis.net/citygml/2.0` | `http://www.opengis.net/citygml/3.0` |

The **bold namespaces** listed in the table above are new in CityGMl v3.0 compared to v2.0.

Note that the XML schemata for CityGML 3.0 are still in active development and subject to change. For the latest XSD files, please refer to the [OGC CityGML 3.0 Encodings GitHub page](https://github.com/opengeospatial/CityGML-3.0Encodings).

### Supported CityGML feature types and elements
The list of supported CityGML v3.0 feature types and elements in the current implementation is shown as follows (note that this list may change in the future):

+ `gml:name`
+ `creationDate` and other similar date/time elements
+ `externalReference`
+ `core:appearance` (both `app:ParameterizedTexture` and `app:X3DMaterial`)
+ `genericAttribute` (`gen:StringAttribute`, `gen:intAttribute`, `gen:doubleAttribute`, `gen:dateAttribute`, `gen:uriAttribute` and `gen:measureAttribute`)
+ `bldg:function`
+ `bldg:roofType`
+ `bldg:heightAboveGround`
+ `BoundarySurfaces` and `BuildingParts`
+ `AbstractConstructiveElement` such as `Window` and `Door` (as `VoidSurface`)
+ `AbstractConstructionSurface` such as `GroundSurface`, `RoofSurface`, `CeilingSurface`, `OuterCeilingSurface`
+ `BuildingFurnitures`
+ `BuildingInstallations`
+ And many more...

For the list of changes please refer to the [change logs](CHANGES.md).

Below are the visualization screenshots taken in FME of the transformed sample dataset [FZKHaus_v3.gml](output/FZKHaus_v3.gml) in 3D and 2D respectively:

 ![FZKHaus in 3D](images/FZKHaus_3D.png "FZKHaus in 3D")
 
 ![FZKHaus in 2D](images/FZKHaus_2D.png "FZKHaus in 2D")

### Other Notes
Currently, only the Building module is supported.

The implementation was tested on a limited number of samples provide in CityGML 3.0 and thus that may not solve all use cases.



CHANGES.md
# Change Log

### 0.6.0 - See [sample file in CityGML 3.0](output/FZKHaus_v3.gml)

FZKHaus in 3D              |  FZKHaus in 2D
:-------------------------:|:-------------------------:
![FZKHaus in 3D](images/FZKHaus_201903_3D.png "FZKHaus in 3D")  |  ![FZKHaus in 2D](images/FZKHaus_201903_2D.png "FZKHaus in 2D")

---------

##### NEW

* **Module `CITYFURNITURE`**:
    * Added transformation for this module

* **Module `BUILDING`**:
    * Added support for ``OuterBuildingInstallation`` and `InteriorBuildingInstallation` in CityGML 2.0 (see [`48b5a7c`](https://github.com/tum-gis/citygml2-to-citygml3/commit/48b5a7c73d0dcc4a4c0dc5a31250d5e6d1b87ad6))

* **Module `BUILDING` and `CITYFURNITURE`**:
    * Added support for ``lodXGeometry`` in CityGML 2.0 with `X` in `{1, 2, 3}` (see [`48b5a7c`](https://github.com/tum-gis/citygml2-to-citygml3/commit/48b5a7c73d0dcc4a4c0dc5a31250d5e6d1b87ad6)):
        * if ``lodXGeometry`` contains a `CompositeSurface`, `Surface`, `Polygon`, `OrientableSurface, TriangularSurface` or `TIN`: 
        it shall be replaced with a corresponding `lodXMultiSurface` with a single `surfaceMember` representing its geometry
        * if ``lodXGeometry`` contains a `Solid` or `CompositeSolid`: it shall be renamed to `lodXSolid`
        * if ``lodXGeometry`` contains a `MultiCurve`: it shall be renamed to `lodXMultiCurve`
        * if ``lodXGeometry`` contains a `MultiSurface`: it shall be renamed to `lodXMultiSurface`
    * All ``lod4Geometry`` elements shall be transformed in the same manner as with other LODs. However, depending on the variable `LOD4_TO_LOD3` in the configuration file [SETTINGS.txt](SETTINGS.txt), `lod4Geometry` elements will be either removed from the dataset or changed to a corresponding LOD3 element
    
##### CHANGES

* **Module `CONSTRUCTION`**:
    * Renamed `boundary` elements to `con:constructionSpace` that are used to link `con:AbstractConstructionSurfaces` (such as `con:WallSurface` and `con:RoofSurface`) with `con:AbstractContruction` elements (such as `bldg:Building`) (see #4 and [`c9f02dd`](https://github.com/tum-gis/citygml2-to-citygml3/commit/c9f02ddc376e1c63b22f0d5fe0497852a3d26a8f))

##### FIXES

* **Module `APPEARANCE`**:
    * Fixed transformation of `app:target` elements (see [`c9f02dd`](https://github.com/tum-gis/citygml2-to-citygml3/commit/c9f02ddc376e1c63b22f0d5fe0497852a3d26a8f))
    
  
### 0.5.0

##### UPDATES

* **Others**:
    * Updated the schemata, transformation and sample CityGML 3.0 datasets accordingly to the new release of [CityGML-3.0CM / 3.0.0-draft.2019.06.05.1](https://github.com/opengeospatial/CityGML-3.0CM/releases/tag/3.0.0-draft.2019.06.05.1) and [CityGML-3.0Encodings / 3.0.0-draft.2019.06.05.1](https://github.com/opengeospatial/CityGML-3.0Encodings/releases/tag/3.0.0-draft.2019.06.05.1) (see [`f821e8b`](https://github.com/tum-gis/citygml2-to-citygml3/commit/f821e8b6575dfd62dcff21604dc5356c27cc5b63)) 

##### FIXES

* **Module `BUILDING`**:
    * Fixed transformation for `BuildingPart` (see [`3fb2b87`](https://github.com/tum-gis/citygml2-to-citygml3/commit/3fb2b87f804c5914317ffabaa454d8c96009de3e) and [`875bc6b`](https://github.com/tum-gis/citygml2-to-citygml3/commit/875bc6bd4e7d7abf81a4a97c93225d15c83883f7))
    
        
### 0.4.0

##### NEW

* **Module `BUILDING`**:
    * Transformation from `bldg:Window` to `con:WindowSurface` and `bldg:Door` to `con:DoorSurface` (see [`67697b8`](https://github.com/tum-gis/citygml2-to-citygml3/commit/67697b8473e6d6858375d49a116dd5da77da283c))
    * Transformation from `bldg:yearOfRenovation` to `con:constructionEvent` (with ``event`` = `endOfRenovation` and `dateOfEvent` = standardized value of `yearOfRenovation`) (see [`3d1adde`](https://github.com/tum-gis/citygml2-to-citygml3/commit/3d1addebce9483bbbb2d33902339d7d2cad45f94))

##### UPDATES

* **Module `APPEARANCE`**:
    * Element `app:surfaceDataMember` changed to `app:surfaceData` (see [`653bc43`](https://github.com/tum-gis/citygml2-to-citygml3/commit/653bc43b9922fcae75cbbaf5f59048d6ee15cf34))
    * Element `app:target` changed to `app:textureParameterization` (see [`ef3c626`](https://github.com/tum-gis/citygml2-to-citygml3/commit/ef3c626d1c54a168687357c9625a6b0794cac7de))
    * Element `app:appearance` changed to `appearance` and `app:target` of `app:TextureAssociation` changed to `app:textureParameterization` (see [`85b6705`](https://github.com/tum-gis/citygml2-to-citygml3/commit/85b6705e4e2cd8106ec23f831147e22facc16088))
    
* **Module `BUILDING`**:
    * Element `bldg:interiorRoom` changed to `bldg:buildingRoom` (see [`177a51a`](https://github.com/tum-gis/citygml2-to-citygml3/commit/177a51a09fc26a043fe1dad479eb43001503986c))
    * Class `bldg:Room` changed to `bldg:BuildingRoom` (see [`3c1dde6`](https://github.com/tum-gis/citygml2-to-citygml3/commit/3c1dde6cd463cc3c064cb2a98c87fddf00141e74))
    * Element `bldg:roomInstallation` changed to `bldg:buildingInstallation` (see [`ef5d7d1`](https://github.com/tum-gis/citygml2-to-citygml3/commit/ef5d7d1d932ff77d7188223fe12d866a5f7d7d68))
    * Add first transformation for element `bldg:buildingSubdivisionInstallation` (see [`8f6eccd`](https://github.com/tum-gis/citygml2-to-citygml3/commit/8f6eccda392d2e54935ddf4d4497b0c0f0ed89a2))
    * Element `bldg:interiorFurniture` changed to `bldg:buildingFurniture` (see [`7b9692b`](https://github.com/tum-gis/citygml2-to-citygml3/commit/7b9692b60277f20dda486eb9087f972715b48baf))
    * Element `bldg:measuredHeight` changed to `con:height` (see [`7d4f0f9`](https://github.com/tum-gis/citygml2-to-citygml3/commit/7d4f0f909c8e8e40617ffee7c2e1a1238e7dee30))
    * Update `bldg:yearOfConstruction` and `bldg:yearOfDemolition` (see [`0f6d02a`](https://github.com/tum-gis/citygml2-to-citygml3/commit/0f6d02af0a55ceb1a10241287d082b165c63b7fc) and [`4576587`](https://github.com/tum-gis/citygml2-to-citygml3/commit/45765877ffe0a60231cbe20c4dab3c363bad3404))

src/Transform.java

/*
 * citygml2-to-citygml3
 *
 * Developer:
 * Son H. Nguyen (son.nguyen@tum.de)
 *
 * With support regarding CityGML encodings from
 * Thomas H. Kolbe (thomas.kolbe@tum.de)
 * Tatjana Kutzner (kutzner@tum.de)
 *
 * MIT License
 *
 * Copyright (c) 2018 Chair of Geoinformatics, Technical University of Munich
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.HashMap;

import javax.xml.transform.Source;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.TransformerFactoryConfigurationError;
import javax.xml.transform.stream.StreamResult;
import javax.xml.transform.stream.StreamSource;

public class Transform {
	public static final String sourceXMLFile = "INPUT_FILE";
	public static final String sourceXSLFile = "XSL_FILE";
	public static final String outputXMLFile = "OUTPUT_FILE";
	public static final String lod4ToLod3 = "LOD4_TO_LOD3";
	public static final String changeLod4Geometry = "LOD4_GEOMETRY";

	public static HashMap<String, String> SETTINGS = new HashMap<>();

	public static void readConfigs(String filename) {
		BufferedReader br = null;
		try {
			br = new BufferedReader(new FileReader(filename));
			String line = "";
			while ((line = br.readLine()) != null) {
				if (line.trim().length() == 0 || line.trim().startsWith("#")) {
					continue;
				}

				String[] ss = line.split("=");
				SETTINGS.put(ss[0], ss[1].replaceAll("\\\\", "/"));
			}
		} catch (Exception e) {
			e.printStackTrace();
		} finally {
			try {
				if (br != null) {
					br.close();
				}
			} catch (IOException e) {
				e.printStackTrace();
			}
		}
	}

	public static void main(String[] args) throws TransformerFactoryConfigurationError, TransformerException, IOException {
		readConfigs("SETTINGS.txt");

		String sourceXMLFile = SETTINGS.get(Transform.sourceXMLFile);
		String sourceXSLFile = SETTINGS.get(Transform.sourceXSLFile);
		String outputXMLFile = SETTINGS.get(Transform.outputXMLFile);
		String lod4ToLod3 = SETTINGS.get(Transform.lod4ToLod3);
		String changeLod4Geometry = SETTINGS.get(Transform.changeLod4Geometry);

		String outputXMLFile_tmp = "output/Tmp.gml";

		System.out.println("Transforming...");
		long startTime = System.nanoTime();

		Source xmlInput = new StreamSource(new File(sourceXMLFile));
		Source xsl = new StreamSource(new File(sourceXSLFile));
		File outputXMLTmp = new File(outputXMLFile_tmp);
		javax.xml.transform.Result xmlOutput = new StreamResult(outputXMLTmp);

		Transformer transformer = TransformerFactory.newInstance().newTransformer(xsl);
		transformer.setParameter("lod4ToLod3", lod4ToLod3);
		transformer.setParameter("changeLod4Geometry", changeLod4Geometry);
		transformer.transform(xmlInput, xmlOutput);

		BufferedReader bReader = new BufferedReader(new java.io.FileReader(outputXMLTmp));
		BufferedWriter bWriter = new BufferedWriter(new FileWriter(new File(outputXMLFile)));

		String line = null;
		boolean found = false;

		String xsdLocation = "http://www.3dcitydb.org/citygml3/2019-06-05/xsds/";

		ArrayList<String[]> ns = new ArrayList<>();
		ns.add(new String[] { "app", "http://www.opengis.net/citygml/appearance/3.0", xsdLocation + "appearance.xsd" });
		ns.add(new String[] { "brid", "http://www.opengis.net/citygml/bridge/3.0", xsdLocation + "bridge.xsd" });
		ns.add(new String[] { "bldg", "http://www.opengis.net/citygml/building/3.0", xsdLocation + "building.xsd" });
		ns.add(new String[] { "frn", "http://www.opengis.net/citygml/cityfurniture/3.0", xsdLocation + "cityFurniture.xsd" });
		ns.add(new String[] { "grp", "http://www.opengis.net/citygml/cityobjectgroup/3.0", xsdLocation + "cityObjectGroup.xsd" });
		ns.add(new String[] { "con", "http://www.opengis.net/citygml/construction/3.0", xsdLocation + "construction.xsd" });
		ns.add(new String[] { "pcl", "http://www.opengis.net/citygml/pointcloud/3.0", xsdLocation + "pointCloud.xsd" });
		ns.add(new String[] { "core", "http://www.opengis.net/citygml/3.0", xsdLocation + "cityGMLBase.xsd" });
		ns.add(new String[] { "dyn", "http://www.opengis.net/citygml/dynamizer/3.0", xsdLocation + "dynamizer.xsd" });
		ns.add(new String[] { "gen", "http://www.opengis.net/citygml/generics/3.0", xsdLocation + "generics.xsd" });
		ns.add(new String[] { "luse", "http://www.opengis.net/citygml/landuse/3.0", xsdLocation + "landUse.xsd" });
		ns.add(new String[] { "dem", "http://www.opengis.net/citygml/relief/3.0", xsdLocation + "relief.xsd" });
		ns.add(new String[] { "tran", "http://www.opengis.net/citygml/transportation/3.0", xsdLocation + "transportation.xsd" });
		ns.add(new String[] { "tun", "http://www.opengis.net/citygml/tunnel/3.0", xsdLocation + "tunnel.xsd" });
		ns.add(new String[] { "veg", "http://www.opengis.net/citygml/vegetation/3.0", xsdLocation + "vegetation.xsd" });
		ns.add(new String[] { "vers", "http://www.opengis.net/citygml/versioning/3.0", xsdLocation + "versioning.xsd" });
		ns.add(new String[] { "wtr", "http://www.opengis.net/citygml/waterbody/3.0", xsdLocation + "waterBody.xsd" });
		ns.add(new String[] { "tsml", "http://www.opengis.net/tsml/1.0", "http://schemas.opengis.net/tsml/1.0/timeseriesML.xsd" });
		ns.add(new String[] { "sos", "http://www.opengis.net/sos/2.0", "http://schemas.opengis.net/sos/2.0/sosGetObservation.xsd" });
		ns.add(new String[] { "xAL", "urn:oasis:names:tc:ciq:xsdschema:xAL:2.0", "http://schemas.opengis.net/citygml/xAL/xAL.xsd" });
		ns.add(new String[] { "xlink", "http://www.w3.org/1999/xlink", "" });
		ns.add(new String[] { "xsi", "http://www.w3.org/2001/XMLSchema-instance", "" });
		ns.add(new String[] { "gml", "http://www.opengis.net/gml/3.2", "" });
		ns.add(new String[] { "ade", "http://www.3dcitydb.org/citygml-ade/3.0/citygml/1.0", "" });
		ns.add(new String[] { "", "http://www.opengis.net/citygml/3.0", "" });

		while ((line = bReader.readLine()) != null) {
			if ((!found) && (line.trim().contains("<CityModel"))) {
				found = true;

				line = "<CityModel gml:id=\"cm1\" ";
				for (int i = 0; i < ns.size(); i++) {
					line += "xmlns" + (ns.get(i)[0].equals("") ? "" : ":") + ns.get(i)[0] + "=\"" + ns.get(i)[1] + "\" ";
				}

				line += "xsi:schemaLocation=\"";
				for (int i = 0; i < ns.size(); i++) {
					line += (ns.get(i)[2].equals("") ? "" : (ns.get(i)[1] + " " + ns.get(i)[2] + " "));
				}

				// Remove the last whitespace
				line = line.substring(0, line.length() - 1);
				line += "\">";
			}

			bWriter.write(line);
			bWriter.newLine();
		}

		bReader.close();
		bWriter.close();
		// Remove temp file
		Files.delete(outputXMLTmp.toPath());

		long endTime = System.nanoTime();
		System.out.println("Transformation done in " + ((long) ((endTime - startTime) / 1e6)) / 1000. + " seconds!\n" + "File output: " + outputXMLFile);
	}
}
